pkg/bpf/collection.go

```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		if prog.SectionName == "tc" || prog.SectionName == "xdp" {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		var id, slot uint32
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			continue
+		}
+
+		const (
+			cilium_policy_map        = 1
+			cilium_calls_map         = 2
+			cilium_egress_policy_map = 4
+		)
+
+		// Any programs in the cilium_policy_map or cilium_egress_policy_map are
+		// called dynamically, assume they are reachable and treat as entrypoints.
+		if id == cilium_policy_map || id == cilium_egress_policy_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		// Any programs not in the tail call map can be ignored (always load them).
```

Comment:
`always load them` confuses me, shouldn't we add them to `entrypoints` then?

Reply:
Hmm. I added this as guard condition I didn't expect to be actually used. But I guess that if someone starts using a new type of section name, and that new program uses tail calls, then we would want to treat it as an entrypoint. That would make this code more future proof. :+1: